### ISSUE TITLE
Add `SourceLocation` for parsing error & debug

### ISSUE THREAD

**Issue by vlopes11** [0xPolygonMiden/miden-vm#857]
Currently, we display a parsing error based on the line number of a compilation. This is introduced by #854

However, this is very simplistic and we might need a better representation that is flexible enough for advanced functionality, such as source mapping.

It is a common standard to use as location a combination of path, line, column. These 3 combined can uniquely map to a string that can be displayed to the user, to give him feedback.

One possibility is to have the following struct:

```rust
pub struct SourceLocation {
    pub path: String,
    pub line: u32,
    pub column: u32,
}
```

However, for the parsing process, this `SourceLocation` would exist for every `Token`. This might become very expensive for large compilations as we would greatly increase the memory requirements with the additional `path`. In fact, this information would be mostly duplicated, as normally we have a couple of files for compilation, and a far greater tokens count than files count. One alternative is to create a bidirectional map `path <-> id`, and use this `id` as part of the source location.

```rust
pub struct SourceLocation {
    pub path_id: usize,
    pub line: u32,
    pub column: u32,
}
```

This would introduce the requirement to have a common context for compilation that would hold the bidirectional map. Such context could look like

```rust
pub struct SourcePathRepository {
    path_ids: BTreeMap<String, usize>,
    paths: Vec<String>,
}
```

The `path_ids` is an efficient lookup to avoid duplicated paths, and `paths` would hold the indexes defined by the ids. This way, we can efficiently append new paths, as we would first lookup the `path_ids` and insert only if it doesn't exist, and we have `O(1)` to return a path, given an id.

There is also another possibility to further improve this context. The `SourceLocation` assumes the path exists, but it might be that it will not always be readily available. This might happen if we compile a source with a bundled, pre-compiled library. We could have a set of predefined sources that will be available for any application that will read these source locations and display them to the user:

```rust
pub enum Source {
    StandardLibrary {
        module: String,
        version: String,
    },
    AbsolutePath(String),
    RawContents(String),
}
```

And replace the simple string by this enum

```rust
pub struct SourcePathRepository {
    path_ids: BTreeMap<Source, usize>,
    paths: Vec<Source>,
}
```

This `SourcePathRepository` can be easily extended into a solution that will support source mapping for debug purposes.

_Originally posted by @bobbinth in https://github.com/0xPolygonMiden/miden-vm/pull/854#discussion_r1167361091_

**Comment by bobbinth** [comment]
I agree with the above, though I think the should start simple and implement `SourceLocation` just as
```Rust
pub struct SourceLocation {
    pub line: u32,
    pub column: u32,
}
```
Once this is working, we can add `path_id` to `SourceLocation`.  In our case, `path` would probably be a logical path to a module (rather than a file path) - e.g., `std::math::u64` - and the mapping of logical paths to files would need to be provided separately.

**Comment by bobbinth** [comment]
Closed by #865 and #861.
